File: backend/services/research.py
"""AI research. Primary provider = Moonshot Kimi (cloud); backup = local Gemma
via Ollama. NO MOCK FALLBACK — if every provider fails the caller raises a 503.

Calls go through providers in order (kimi -> ollama). Robustly parses the
returned JSON (strips code fences). The returned dict's ``provider``/``model``
reflect whoever actually answered.
"""
from __future__ import annotations

import json
import re
from datetime import datetime, timezone
from typing import Any

import httpx

from config import logger, settings
from services import alpaca_service, indicators, kimi

_OLLAMA_TIMEOUT = 120.0


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def ollama_connected() -> bool:
    try:
        r = httpx.get(f"{settings.ollama_base_url}/api/tags", timeout=3.0)
        return r.status_code == 200
    except Exception:
        return False


def _provider_order() -> list[str]:
    """Ordered list of distinct providers to try: primary then backup."""
    out: list[str] = []
    for p in (settings.research_provider, settings.research_backup_provider):
        p = (p or "").lower().strip()
        if p and p not in out:
            out.append(p)
    return out or ["ollama"]


def _call_provider(provider: str, system: str, user: str, *,
                   json_format: bool, num_predict: int) -> tuple[str, str]:
    """Call one provider. Returns (content, model_id). Raises on failure."""
    if provider == "kimi":
        content = kimi.chat(system, user, json_format=json_format, max_tokens=max(num_predict, 1500))
        return content, settings.kimi_model
    if provider == "ollama":
        content = _ollama_chat(system, user, json_format=json_format, num_predict=num_predict)
        return content, settings.research_model
    raise RuntimeError(f"unknown research provider '{provider}'")


_SYSTEM = (
    "You are a disciplined equity analyst. You always respond with a single STRICT "
    "JSON object and nothing else. Keys: thesis (string), sentiment_score (number in "
    "[-1,1]), conviction (number in [0,100]), key_risks (array of strings), "
    "suggested_action (one of buy/hold/reduce/sell), suggested_stop (number, a price), "
    "suggested_target (number, a price), regime (one of trend-up/trend-down/range/"
    "high-vol), bear_case (string articulating the strongest opposing case)."
)


def _build_prompt(symbol: str, snapshot: dict, ind: dict, bars: list[dict],
                  news: list[dict]) -> str:
    recent_bars = bars[-20:]
    bar_lines = "\n".join(
        f"{b.get('t', '')[:10]} O{b['o']} H{b['h']} L{b['l']} C{b['c']} V{b['v']}"
        for b in recent_bars
    )
    news_lines = "\n".join(f"- {n['headline']} ({n['source']})" for n in news)
    return (
        f"Analyze {symbol}.\n\n"
        f"Latest snapshot: {json.dumps(snapshot)}\n\n"
        f"Indicator snapshot: {json.dumps(ind)}\n\n"
        f"Recent daily bars (oldest->newest):\n{bar_lines}\n\n"
        f"Recent headlines:\n{news_lines or '(none)'}\n\n"
        "Produce a concise trading thesis, then reason through the strongest bear_case. "
        "Give a sentiment_score in [-1,1], a conviction in [0,100], concrete "
        "suggested_stop and suggested_target price levels near the current price, a "
        "suggested_action, and the current market regime. Return STRICT JSON only."
    )


def _parse_json_loose(text: str) -> dict[str, Any]:
    text = (text or "").strip()
    # strip ```json ... ``` fences if present
    fence = re.match(r"^```(?:json)?\s*(.*?)\s*```$", text, re.DOTALL)
    if fence:
        text = fence.group(1).strip()
    try:
        return json.loads(text)
    except Exception:
        # last-ditch: grab the first {...} block
        m = re.search(r"\{.*\}", text, re.DOTALL)
        if m:
            return json.loads(m.group(0))
        raise


def _coerce(data: dict[str, Any], symbol: str, snapshot: dict, provider: str,
            model: str) -> dict[str, Any]:
    price = snapshot.get("price", 100.0) or 100.0

    def num(v, default):
        try:
            return float(v)
        except (TypeError, ValueError):
            return default

    risks = data.get("key_risks") or []
    if isinstance(risks, str):
        risks = [risks]
    return {
        "symbol": symbol,
        "thesis": str(data.get("thesis", "")).strip(),
        "sentiment_score": max(-1.0, min(1.0, num(data.get("sentiment_score"), 0.0))),
        "conviction": max(0.0, min(100.0, num(data.get("conviction"), 50.0))),
        "key_risks": [str(r) for r in risks][:8],
        "suggested_action": str(data.get("suggested_action", "hold")).lower(),
        "suggested_stop": num(data.get("suggested_stop"), round(price * 0.95, 2)),
        "suggested_target": num(data.get("suggested_target"), round(price * 1.08, 2)),
        "regime": str(data.get("regime", "range")),
        "bear_case": str(data.get("bear_case", "")).strip(),
        "provider": provider,
        "model": model,
        "generated_at": _now(),
    }


class ResearchUnavailable(Exception):
    """Raised when every research provider fails. Routers map this to a 503."""


def _ollama_chat(
    system: str,
    user: str,
    *,
    json_format: bool = True,
    num_predict: int = 700,
) -> str:
    payload = {
        "model": settings.research_model,
        "stream": False,
        # gemma4 is a reasoning model; thinking tokens otherwise starve the
        # actual output and we get empty content. Disable it.
        "think": False,
        # Keep the model resident so subsequent calls don't pay reload latency.
        "keep_alive": "30m",
        "options": {"num_predict": num_predict, "temperature": 0.4},
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
    }
    if json_format:
        payload["format"] = "json"
    with httpx.Client(timeout=_OLLAMA_TIMEOUT) as client:
        r = client.post(f"{settings.ollama_base_url}/api/chat", json=payload)
        r.raise_for_status()
        return r.json()["message"]["content"]


def _safe_news(symbols: list[str], limit: int) -> list[dict[str, Any]]:
    """News is supplementary context — never let its absence block research."""
    try:
        return alpaca_service.get_news(symbols, limit)
    except Exception as exc:
        logger.info("research: news unavailable for %s (%s) — proceeding without.",
                    symbols, exc)
        return []


def analyze(symbol: str) -> dict[str, Any]:
    symbol = symbol.upper()
    bars = alpaca_service.get_bars(symbol, "1Day", 120)
    ind = indicators.compute_all(bars)
    snapshot = alpaca_service.get_snapshot(symbol)
    news = _safe_news([symbol], 5)
    prompt = _build_prompt(symbol, snapshot, ind, bars, news)

    last_err: Exception | None = None
    for provider in _provider_order():
        try:
            content, model = _call_provider(
                provider, _SYSTEM, prompt, json_format=True, num_predict=900)
            data = _parse_json_loose(content)
            return _coerce(data, symbol, snapshot, provider, model)
        except Exception as exc:
            last_err = exc
            logger.warning("research analyze via %s failed (%s) — trying next provider.",
                           provider, exc)
    raise ResearchUnavailable(f"all research providers failed: {last_err}")


def briefing(watchlist: list[str]) -> dict[str, Any]:
    snapshots = alpaca_service.get_snapshots(watchlist)
    items = []
    for sym in watchlist:
        snap = snapshots.get(sym, {})
        chg = snap.get("change_pct", 0.0)
        sentiment = max(-1.0, min(1.0, chg / 5.0))
        note = (
            f"{sym} {'up' if chg >= 0 else 'down'} {abs(chg):.2f}% "
            f"at ${snap.get('price', 0):.2f}."
        )
        items.append({"symbol": sym, "note": note, "sentiment": round(sentiment, 3)})

    avg = sum(i["sentiment"] for i in items) / max(1, len(items))
    regime = "trend-up" if avg > 0.15 else ("trend-down" if avg < -0.15 else "range")
    summary = (
        f"Watchlist breadth is {'positive' if avg > 0 else 'mixed'} this session. "
        f"Average sentiment {avg:+.2f}; regime reads {regime}."
    )

    user = (
        "Write a concise 2-sentence morning market briefing for this watchlist "
        f"snapshot: {json.dumps(items)}. Be concrete. Respond as JSON {{\"summary\": string}}."
    )
    for provider in _provider_order():
        try:
            content, _ = _call_provider(
                provider, "You are a market strategist. Respond with STRICT JSON only.",
                user, json_format=True, num_predict=400)
            summary = _parse_json_loose(content).get("summary", summary)
            break
        except Exception as exc:
            logger.warning("briefing via %s failed (%s) — trying next.", provider, exc)

    return {
        "generated_at": _now(),
        "summary": summary,
        "items": items,
        "regime": regime,
    }


_DEEP_SYSTEM = (
    "You are a senior equity research analyst writing an internal markdown report. "
    "Write clear, well-structured GitHub-flavored markdown with section headers (##), "
    "bullet points, and concrete numbers drawn from the supplied data. Do NOT invent "
    "specific figures you were not given. Be decisive but balanced."
)


def _deep_disclaimer(kind: str) -> str:
    if kind == "earnings":
        return (
            "> _Note: earnings detail below is LLM-synthesized from recent price action "
            "and available news headlines — it is NOT a verified earnings transcript or "
            "fundamentals feed. Treat dates/figures as indicative only._\n\n"
        )
    return (
        "> _Note: this report is LLM-synthesized from market data and news headlines "
        "available to the app. Verify any specific claims independently._\n\n"
    )


def generate_deep(symbol: str, kind: str = "deep") -> dict[str, Any]:
    """Build a longer markdown research report (deep dive or earnings synthesis)."""
    symbol = symbol.upper()
    kind = kind if kind in ("deep", "earnings") else "deep"
    bars = alpaca_service.get_bars(symbol, "1Day", 120)
    ind = indicators.compute_all(bars)
    snapshot = alpaca_service.get_snapshot(symbol)
    news = _safe_news([symbol], 8)

    recent_bars = bars[-30:]
    bar_lines = "\n".join(
        f"{b.get('t', '')[:10]} O{b['o']} H{b['h']} L{b['l']} C{b['c']} V{b['v']}"
        for b in recent_bars
    )
    news_lines = "\n".join(f"- {n['headline']} ({n.get('source','')})" for n in news)

    if kind == "earnings":
        ask = (
            f"Write an EARNINGS-FOCUSED markdown report for {symbol}. Sections: "
            "## Setup into earnings, ## What the market expects, ## Key items to watch, "
            "## Bull case, ## Bear case, ## Options-implied move (qualitative), "
            "## Trade plan. Be explicit that earnings specifics are synthesized from "
            "news, not a verified transcript."
        )
    else:
        ask = (
            f"Write a DEEP-DIVE markdown research report for {symbol}. Sections: "
            "## Overview, ## Technical picture, ## Momentum & trend, ## Catalysts & news, "
            "## Bull case, ## Bear case, ## Risks, ## Trade plan (entries, stops, targets)."
        )

    user = (
        f"{ask}\n\n"
        f"Latest snapshot: {json.dumps(snapshot)}\n\n"
        f"Indicator snapshot: {json.dumps(ind)}\n\n"
        f"Recent daily bars (oldest->newest):\n{bar_lines}\n\n"
        f"Recent headlines:\n{news_lines or '(none)'}\n\n"
        "Return ONLY markdown (no JSON, no preamble)."
    )

    provider = None
    model = None
    body = ""
    last_err: Exception | None = None
    for prov in _provider_order():
        try:
            content, mdl = _call_provider(
                prov, _DEEP_SYSTEM, user, json_format=False, num_predict=1400)
            content = (content or "").strip()
            if content:
                provider, model, body = prov, mdl, content
                break
        except Exception as exc:
            last_err = exc
            logger.warning("generate_deep via %s failed for %s (%s) — trying next.",
                           prov, symbol, exc)

    if not body:
        raise ResearchUnavailable(
            f"deep research unavailable for {symbol}: {last_err or 'empty response'}")

    title = f"{symbol} {'earnings preview' if kind == 'earnings' else 'deep dive'}"
    full_body = _deep_disclaimer(kind) + body
    return {
        "symbol": symbol,
        "kind": kind,
        "title": title,
        "body": full_body,
        "data": {
            "snapshot": snapshot,
            "indicators": ind,
            "news": news[:8],
        },
        "provider": provider,
        "model": model,
        "generated_at": _now(),
    }


def regime() -> dict[str, Any]:
    spy = alpaca_service.get_snapshot("SPY")
    qqq = alpaca_service.get_snapshot("QQQ")
    spy_chg = spy.get("change_pct", 0.0)
    qqq_chg = qqq.get("change_pct", 0.0)
    avg = (spy_chg + qqq_chg) / 2

    bars = alpaca_service.get_bars("SPY", "1Day", 60)
    ind = indicators.compute_all(bars)
    atr = ind.get("atr14") or 0.0
    price = spy.get("price", 1.0) or 1.0
    vix_proxy = round((atr / price) * 100 * 16, 2)

    if vix_proxy > 25:
        reg = "high-vol"
    elif avg > 0.3:
        reg = "trend-up"
    elif avg < -0.3:
        reg = "trend-down"
    else:
        reg = "range"

    breadth = round(max(0.0, min(1.0, 0.5 + avg / 4)), 3)
    return {
        "regime": reg,
        "vix_proxy": vix_proxy,
        "breadth": breadth,
        "note": (
            f"SPY {spy_chg:+.2f}% / QQQ {qqq_chg:+.2f}%. Volatility proxy {vix_proxy}. "
            f"Regime: {reg}."
        ),
    }

File: backend-rs/src/research.rs
//! AI research service. Kimi primary -> Gemma backup. NO MOCK FALLBACK.

use crate::indicators;
use crate::llm::parse_json_loose;
use crate::state::AppState;
use chrono::Utc;
use serde_json::{json, Value};

fn now() -> String {
    Utc::now().to_rfc3339()
}

const SYSTEM: &str = "You are a disciplined equity analyst. You always respond with a single STRICT JSON object and nothing else. Keys: thesis (string), sentiment_score (number in [-1,1]), conviction (number in [0,100]), key_risks (array of strings), suggested_action (one of buy/hold/reduce/sell), suggested_stop (number, a price), suggested_target (number, a price), regime (one of trend-up/trend-down/range/high-vol), bear_case (string articulating the strongest opposing case).";

/// Ordered list of distinct providers: primary then backup.
fn provider_order(state: &AppState) -> Vec<String> {
    let mut out = vec![];
    for p in [
        &state.settings.research_provider,
        &state.settings.research_backup_provider,
    ] {
        let p = p.to_lowercase();
        let p = normalize_provider(Some(&p)).unwrap_or(p);
        if !p.is_empty() && !out.contains(&p) {
            out.push(p);
        }
    }
    if out.is_empty() {
        out.push("ollama".into());
    }
    out
}

pub fn normalize_provider(provider: Option<&str>) -> Option<String> {
    let p = provider?.to_lowercase();
    let p = p.trim();
    if p == "gemma" {
        Some("ollama".into())
    } else {
        Some(p.to_string())
    }
}

fn depth_preset(depth: &str) -> (usize, usize, usize, u32) {
    // (bars, recent_bars, news, num_predict)
    match depth {
        "quick" => (40, 10, 0, 400),
        "deep" => (120, 40, 8, 1200),
        _ => (120, 20, 5, 900),
    }
}

async fn call_provider(
    state: &AppState,
    provider: &str,
    system: &str,
    user: &str,
    json_format: bool,
    num_predict: u32,
) -> anyhow::Result<(String, String)> {
    match provider {
        "kimi" => {
            let content = state
                .llm
                .kimi_chat(system, user, json_format, num_predict.max(1500))
                .await?;
            Ok((content, state.settings.kimi_model.clone()))
        }
        "ollama" => {
            let content = state
                .llm
                .ollama_chat(system, user, json_format, num_predict)
                .await?;
            Ok((content, state.settings.research_model.clone()))
        }
        other => anyhow::bail!("unknown research provider '{}'", other),
    }
}

fn build_prompt(symbol: &str, snapshot: &Value, ind: &Value, bars: &[Value], news: &[Value], depth: &str) -> String {
    let (_, recent_n, _, _) = depth_preset(depth);
    let recent: Vec<&Value> = bars.iter().rev().take(recent_n).rev().collect();
    let bar_lines: String = recent
        .iter()
        .map(|b| {
            let t = b["t"].as_str().unwrap_or("");
            let t = if t.len() >= 10 { &t[..10] } else { t };
            format!(
                "{} O{} H{} L{} C{} V{}",
                t, b["o"], b["h"], b["l"], b["c"], b["v"]
            )
        })
        .collect::<Vec<_>>()
        .join("\n");

    if depth == "quick" {
        return format!(
            "Analyze {} quickly.\n\nLatest snapshot: {}\n\nIndicator snapshot: {}\n\nRecent daily bars (oldest->newest):\n{}\n\nGive a brief trading thesis, a short bear_case, sentiment_score in [-1,1], conviction in [0,100], suggested_stop and suggested_target near the current price, a suggested_action, and the market regime. STRICT JSON only.",
            symbol, snapshot, ind, bar_lines
        );
    }
    let news_lines: String = news
        .iter()
        .map(|n| format!("- {} ({})", n["headline"].as_str().unwrap_or(""), n["source"].as_str().unwrap_or("")))
        .collect::<Vec<_>>()
        .join("\n");
    format!(
        "Analyze {}.\n\nLatest snapshot: {}\n\nIndicator snapshot: {}\n\nRecent daily bars (oldest->newest):\n{}\n\nRecent headlines:\n{}\n\nProduce a concise trading thesis, then reason through the strongest bear_case. Give a sentiment_score in [-1,1], a conviction in [0,100], concrete suggested_stop and suggested_target price levels near the current price, a suggested_action, and the current market regime. Return STRICT JSON only.",
        symbol, snapshot, ind, bar_lines, if news_lines.is_empty() { "(none)".to_string() } else { news_lines }
    )
}

fn num(v: &Value, default: f64) -> f64 {
    match v {
        Value::Number(n) => n.as_f64().unwrap_or(default),
        Value::String(s) => s.parse().unwrap_or(default),
        _ => default,
    }
}

fn coerce(data: &Value, symbol: &str, snapshot: &Value, provider: &str, model: &str) -> Value {
    let price = num(&snapshot["price"], 0.0);
    let mut risks: Vec<Value> = match &data["key_risks"] {
        Value::Array(a) => a.iter().map(|r| json!(r.as_str().map(|s| s.to_string()).unwrap_or_else(|| r.to_string()))).collect(),
        Value::String(s) => vec![json!(s)],
        _ => vec![],
    };
    risks.truncate(8);

    // NO MOCK DATA: do not invent missing fields. Track what the model omitted so
    // consumers (e.g. the bot AI gate) can fail-safe instead of trusting fabricated
    // confidence. Conviction defaults to 0 (gate will not pass on it), and
    // stop/target stay null unless the model actually provided them.
    let mut degraded: Vec<&str> = vec![];
    let thesis = data["thesis"].as_str().unwrap_or("").trim().to_string();
    if thesis.is_empty() {
        degraded.push("thesis");
    }
    let sentiment = if data["sentiment_score"].is_null() {
        degraded.push("sentiment_score");
        0.0
    } else {
        num(&data["sentiment_score"], 0.0).clamp(-1.0, 1.0)
    };
    let conviction = if data["conviction"].is_null() {
        degraded.push("conviction");
        0.0
    } else {
        num(&data["conviction"], 0.0).clamp(0.0, 100.0)
    };
    // Only carry a stop/target the model actually returned (and only if we know price).
    let suggested_stop = if data["suggested_stop"].is_null() {
        Value::Null
    } else {
        json!((num(&data["suggested_stop"], 0.0) * 100.0).round() / 100.0)
    };
    let suggested_target = if data["suggested_target"].is_null() {
        Value::Null
    } else {
        json!((num(&data["suggested_target"], 0.0) * 100.0).round() / 100.0)
    };
    let _ = price;

    json!({
        "symbol": symbol,
        "thesis": thesis,
        "sentiment_score": sentiment,
        "conviction": conviction,
        "key_risks": risks,
        "suggested_action": data["suggested_action"].as_str().unwrap_or("hold").to_lowercase(),
        "suggested_stop": suggested_stop,
        "suggested_target": suggested_target,
        "regime": data["regime"].as_str().unwrap_or("unknown"),
        "bear_case": data["bear_case"].as_str().unwrap_or("").trim(),
        "provider": provider,
        "model": model,
        "degraded": !degraded.is_empty(),
        "degraded_fields": degraded,
        "generated_at": now(),
    })
}

async fn safe_news(state: &AppState, symbols: &[String], limit: usize) -> Vec<Value> {
    if limit == 0 {
        return vec![];
    }
    match state.alpaca.get_news(symbols, limit).await {
        Ok(v) => v.as_array().cloned().unwrap_or_default(),
        Err(_) => vec![],
    }
}

pub async fn analyze(
    state: &AppState,
    symbol: &str,
    provider: Option<&str>,
    depth: &str,
) -> anyhow::Result<Value> {
    let symbol = symbol.to_uppercase();
    let depth = if depth.is_empty() { "standard" } else { depth };
    let (bars_n, _, news_n, num_predict) = depth_preset(depth);

    let forced = normalize_provider(provider);
    let providers = match &forced {
        Some(p) => vec![p.clone()],
        None => provider_order(state),
    };

    let bars_v = state.alpaca.get_bars(&symbol, "1Day", bars_n).await?;
    let bars = bars_v.as_array().cloned().unwrap_or_default();
    let ind = indicators::compute_all(&bars);
    let snapshot = state.alpaca.get_snapshot(&symbol).await?;
    let news = safe_news(state, &[symbol.clone()], news_n).await;
    let prompt = build_prompt(&symbol, &snapshot, &ind, &bars, &news, depth);

    let mut last_err = String::new();
    for prov in &providers {
        match call_provider(state, prov, SYSTEM, &prompt, true, num_predict).await {
            Ok((content, model)) => match parse_json_loose(&content) {
                Ok(data) => {
                    let mut result = coerce(&data, &symbol, &snapshot, prov, &model);
                    // Reject a hollow analysis (no thesis) — try the next provider
                    // rather than returning fabricated/empty content as real.
                    if result["thesis"].as_str().unwrap_or("").trim().is_empty() {
                        last_err = format!("{prov} returned an empty/invalid analysis");
                        continue;
                    }
                    result["depth"] = json!(depth);
                    return Ok(result);
                }
                Err(e) => last_err = e.to_string(),
            },
            Err(e) => last_err = e.to_string(),
        }
    }
    anyhow::bail!("research analyze failed for {}: {}", symbol, last_err)
}

pub async fn briefing(state: &AppState, watchlist: &[String], provider: Option<&str>) -> anyhow::Result<Value> {
    let forced = normalize_provider(provider);
    let snapshots = state.alpaca.get_snapshots(watchlist).await?;
    let mut items = vec![];
    for sym in watchlist {
        let snap = &snapshots[sym];
        let chg = num(&snap["change_pct"], 0.0);
        let sentiment = (chg / 5.0).clamp(-1.0, 1.0);
        let note = format!(
            "{} {} {:.2}% at ${:.2}.",
            sym,
            if chg >= 0.0 { "up" } else { "down" },
            chg.abs(),
            num(&snap["price"], 0.0)
        );
        items.push(json!({"symbol": sym, "note": note, "sentiment": (sentiment * 1000.0).round() / 1000.0}));
    }
    let avg = if items.is_empty() {
        0.0
    } else {
        items.iter().map(|i| num(&i["sentiment"], 0.0)).sum::<f64>() / items.len() as f64
    };
    let regime = if avg > 0.15 {
        "trend-up"
    } else if avg < -0.15 {
        "trend-down"
    } else {
        "range"
    };
    let mut summary = format!(
        "Watchlist breadth is {} this session. Average sentiment {:+.2}; regime reads {}.",
        if avg > 0.0 { "positive" } else { "mixed" },
        avg,
        regime
    );
    let user = format!(
        "Write a concise 2-sentence morning market briefing for this watchlist snapshot: {}. Be concrete. Respond as JSON {{\"summary\": string}}.",
        Value::Array(items.clone())
    );
    let providers = match &forced {
        Some(p) => vec![p.clone()],
        None => provider_order(state),
    };
    for prov in &providers {
        if let Ok((content, _)) = call_provider(
            state,
            prov,
            "You are a market strategist. Respond with STRICT JSON only.",
            &user,
            true,
            400,
        )
        .await
        {
            if let Ok(parsed) = parse_json_loose(&content) {
                if let Some(s) = parsed["summary"].as_str() {
                    summary = s.to_string();
                    break;
                }
            }
        }
    }
    Ok(json!({
        "generated_at": now(),
        "summary": summary,
        "items": items,
        "regime": regime,
    }))
}

const DEEP_SYSTEM: &str = "You are a senior equity research analyst writing an internal markdown report. Write clear, well-structured GitHub-flavored markdown with section headers (##), bullet points, and concrete numbers drawn from the supplied data. Do NOT invent specific figures you were not given. Be decisive but balanced.";

fn deep_disclaimer(kind: &str) -> String {
    if kind == "earnings" {
        "> _Note: earnings detail below is LLM-synthesized from recent price action and available news headlines — it is NOT a verified earnings transcript or fundamentals feed. Treat dates/figures as indicative only._\n\n".to_string()
    } else {
        "> _Note: this report is LLM-synthesized from market data and news headlines available to the app. Verify any specific claims independently._\n\n".to_string()
    }
}

pub async fn generate_deep(state: &AppState, symbol: &str, kind: &str, provider: Option<&str>) -> anyhow::Result<Value> {
    let symbol = symbol.to_uppercase();
    let kind = if kind == "earnings" { "earnings" } else { "deep" };
    let forced = normalize_provider(provider);
    let bars_v = state.alpaca.get_bars(&symbol, "1Day", 120).await?;
    let bars = bars_v.as_array().cloned().unwrap_or_default();
    let ind = indicators::compute_all(&bars);
    let snapshot = state.alpaca.get_snapshot(&symbol).await?;
    let news = safe_news(state, &[symbol.clone()], 8).await;

    let recent: Vec<&Value> = bars.iter().rev().take(30).rev().collect();
    let bar_lines: String = recent
        .iter()
        .map(|b| {
            let t = b["t"].as_str().unwrap_or("");
            let t = if t.len() >= 10 { &t[..10] } else { t };
            format!("{} O{} H{} L{} C{} V{}", t, b["o"], b["h"], b["l"], b["c"], b["v"])
        })
        .collect::<Vec<_>>()
        .join("\n");
    let news_lines: String = news
        .iter()
        .map(|n| format!("- {} ({})", n["headline"].as_str().unwrap_or(""), n["source"].as_str().unwrap_or("")))
        .collect::<Vec<_>>()
        .join("\n");

    let ask = if kind == "earnings" {
        format!("Write an EARNINGS-FOCUSED markdown report for {}. Sections: ## Setup into earnings, ## What the market expects, ## Key items to watch, ## Bull case, ## Bear case, ## Options-implied move (qualitative), ## Trade plan. Be explicit that earnings specifics are synthesized from news, not a verified transcript.", symbol)
    } else {
        format!("Write a DEEP-DIVE markdown research report for {}. Sections: ## Overview, ## Technical picture, ## Momentum & trend, ## Catalysts & news, ## Bull case, ## Bear case, ## Risks, ## Trade plan (entries, stops, targets).", symbol)
    };
    let user = format!(
        "{}\n\nLatest snapshot: {}\n\nIndicator snapshot: {}\n\nRecent daily bars (oldest->newest):\n{}\n\nRecent headlines:\n{}\n\nReturn ONLY markdown (no JSON, no preamble).",
        ask, snapshot, ind, bar_lines, if news_lines.is_empty() { "(none)".to_string() } else { news_lines }
    );

    let providers = match &forced {
        Some(p) => vec![p.clone()],
        None => provider_order(state),
    };
    let mut provider_used = None;
    let mut model = None;
    let mut body = String::new();
    let mut last_err = String::new();
    for prov in &providers {
        match call_provider(state, prov, DEEP_SYSTEM, &user, false, 1400).await {
            Ok((content, mdl)) => {
                let content = content.trim().to_string();
                if !content.is_empty() {
                    provider_used = Some(prov.clone());
                    model = Some(mdl);
                    body = content;
                    break;
                }
            }
            Err(e) => last_err = e.to_string(),
        }
    }
    if body.is_empty() {
        anyhow::bail!("deep research unavailable for {}: {}", symbol, last_err);
    }
    let title = format!("{} {}", symbol, if kind == "earnings" { "earnings preview" } else { "deep dive" });
    let full_body = format!("{}{}", deep_disclaimer(kind), body);
    let news_slice: Vec<Value> = news.iter().take(8).cloned().collect();
    Ok(json!({
        "symbol": symbol,
        "kind": kind,
        "title": title,
        "body": full_body,
        "data": {"snapshot": snapshot, "indicators": ind, "news": news_slice},
        "provider": provider_used,
        "model": model,
        "generated_at": now(),
    }))
}

pub async fn regime(state: &AppState) -> anyhow::Result<Value> {
    let spy = state.alpaca.get_snapshot("SPY").await?;
    let qqq = state.alpaca.get_snapshot("QQQ").await?;
    let spy_chg = num(&spy["change_pct"], 0.0);
    let qqq_chg = num(&qqq["change_pct"], 0.0);
    let avg = (spy_chg + qqq_chg) / 2.0;

    let bars_v = state.alpaca.get_bars("SPY", "1Day", 60).await?;
    let bars = bars_v.as_array().cloned().unwrap_or_default();
    let ind = indicators::compute_all(&bars);
    let atr = num(&ind["atr14"], 0.0);
    let price = {
        let p = num(&spy["price"], 1.0);
        if p == 0.0 {
            1.0
        } else {
            p
        }
    };
    let vix_proxy = ((atr / price) * 100.0 * 16.0 * 100.0).round() / 100.0;
    let reg = if vix_proxy > 25.0 {
        "high-vol"
    } else if avg > 0.3 {
        "trend-up"
    } else if avg < -0.3 {
        "trend-down"
    } else {
        "range"
    };
    let breadth = (0.5 + avg / 4.0).clamp(0.0, 1.0);
    let breadth = (breadth * 1000.0).round() / 1000.0;
    Ok(json!({
        "regime": reg,
        "vix_proxy": vix_proxy,
        "breadth": breadth,
        "note": format!("SPY {:+.2}% / QQQ {:+.2}%. Volatility proxy {}. Regime: {}.", spy_chg, qqq_chg, vix_proxy, reg),
    }))
}
